### DISCUSSION TITLE
Mdx plugin not working as expected with Mdx files in /src/pages

### DISCUSSION THREAD

**Discussion by nicolls1** [gatsbyjs/gatsby#28017] [Help] [answered]
## Description

I'm going through the tutorial here: https://www.gatsbyjs.com/docs/mdx/programmatically-creating-pages/ about making mdx pages and the example uses a folder outside of the `src/pages` directory. I would like to have mdx pages rendered in their normal path inside the directory like is done here: https://www.gatsbyjs.com/tutorial/part-seven/ for md files but everything I try doesn't seem to work. 

### Steps to reproduce

I have a file `src/pages/test.mdx` with contents:
```
---
title: Blog Post
---

Trying out MDX
```

I have a template file with the contents:
```
import React from "react"
import { graphql } from "gatsby"
import { MDXProvider } from "@mdx-js/react"
import { MDXRenderer } from "gatsby-plugin-mdx"
import { Link } from "gatsby"
const shortcodes = { Link } // Provide common components here
export default function PageTemplate({ data: { mdx } }) {
  return (
    <div>
      <h1>{mdx.frontmatter.title}</h1>
      <MDXProvider components={shortcodes}>
        <MDXRenderer>{mdx.body}</MDXRenderer>
      </MDXProvider>
    </div>
  )
}
```

I have set gatsby-source-filesystem to be the normal pages directory: 
```
    {
      resolve: "gatsby-source-filesystem",
      options: {
        name: "pages",
        path: `${__dirname}/src/pages/`,
      },
    },
 ``` 
When I then create a node field like:
```
exports.onCreateNode = ({ node, actions, getNode }) => {
  const { createNodeField } = actions
  if (node.internal.type === "Mdx") {
    const value = createFilePath({ node, getNode })
    createNodeField({
      node,
      name: "slug",
      value: `${value}`,
    })
```
and then use the `exports.createPages` from the tutorial with my own custom component, I notice that my component wrapper is not used and there is no header. If I make an update to the page mdx file I then see the header show up. From what I have seen it looks like the mdx files inside of `src/pages` get added twice to the build process. I verified this by changing the `onCreateNode` to:
```
exports.onCreateNode = ({ node, actions, getNode }) => {
  const { createNodeField } = actions
  if (node.internal.type === "Mdx") {
    const value = createFilePath({ node, getNode })
    createNodeField({
      node,
      name: "slug",
      value: `/subpath${value}`,
    })
```
When I do this, I see a page at `localhost:8000/test/` which has just the body text and then at `localhost:8000/subpath/test/` I have a page with both header and body text. 

I found the docs here: https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-mdx#default-layouts about default layouts for mdx files and when I change the default layout it is updated in the page at `localhost:8000/test/`. The problem is that I would like to pass the node id as context to the page template for a GraphQL query as done in the tutorial page but I don't see a way to do this with the defaultLayouts config option.

Another bit of information( but I think is just a side effect of the file getting overwritten because the same mdx file is being processed twice), I found this command: `rm -r .cache/ ; env "NODE_ENV=development" "DEBUG=gatsby:query-watcher" npm run dev` in a different issue and when I run I get the following: 
```
  gatsby:query-watcher /<DIR>/src/templates/DefaultPageTemplate/DefaultPageTemplate.tsx component was removed because it isn't used by any page +0ms
success extract queries from components - 0.167s
warning The GraphQL query in the non-page component "/<DIR>/src/templates/DefaultPageTemplate/DefaultPageTemplate.tsx" will not be run.
Exported queries are only executed for Page components. It's possible you're
trying to create pages in your gatsby-node.js and that's failing for some
reason.

If the failing component(s) is a regular component and not intended to be a page
component, you generally want to use a <StaticQuery> (https://gatsbyjs.org/docs/static-query)
instead of exporting a page query.

If you're more experienced with GraphQL, you can also export GraphQL
fragments from components and compose the fragments in the Page component
query and pass data down into the child component — https://graphql.org/learn/queries/#fragments
```
It seems the template is getting registered then unregistered and then when its trying to be used later it gives a warning that it is being used outside the page concept. 

Thanks for any help!!

### Expected result

I would like one page from my mdx file with the component I specify in `createPages`.

### Actual result

2 pages get created and I can't figure out how to have just 1 with my desired template and context passed.

### Environment

  System:
    OS: macOS 10.15.7
    CPU: (8) x64 Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 10.16.3 - ~/.nvm/versions/node/v10.16.3/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.9.0 - ~/.nvm/versions/node/v10.16.3/bin/npm
  Languages:
    Python: 2.7.18 - /opt/miniconda2/bin/python
  Browsers:
    Chrome: 86.0.4240.198
    Safari: 14.0
  npmPackages:
    gatsby: ^2.25.3 => 2.25.4
    gatsby-plugin-mdx: ^1.3.1 => 1.3.1
    gatsby-plugin-styled-components: ^3.3.14 => 3.3.14
    gatsby-source-filesystem: ^2.4.2 => 2.4.2
  npmGlobalPackages:
    gatsby-cli: 2.12.117

**Comment by LekoArts** [comment, marked as answer]
Hi!

Have you found this doc page? https://www.gatsbyjs.com/docs/mdx/writing-pages/

`gatsby-plugin-mdx` automatically looks for files inside `src/pages`, so you can remove the gatsby-node.js changes and gatsby-config.js changes -- creating files there should already "just" work.

**Reply by nicolls1** [reply]
Hi! thanks for the quick response. So I was able to fix the double rendering problem by removing the gatsby-node.js changes and now I am running into what I was running into before as well. I cannot use the GraphQL query in the template to get the mdx data. I'm pretty new to gatsby so I think there is just a small detail here I'm missing. 

Here is what I have:
gatsby-config.js:
```
...
  plugins: [
    "gatsby-plugin-styled-components",
    {
      resolve: "gatsby-plugin-mdx",
      options: {
        defaultLayouts: {
          default: path.resolve(`./src/templates/DefaultPageTemplate/DefaultPageTemplate.tsx`),
        }
    }
...
```

`./src/templates/DefaultPageTemplate/DefaultPageTemplate.tsx`:
```
...
const PageTemplate: React.FC<PageTemplateProps> = (props) => {
  console.log(props)
  return (
    <>
      <h2>Header</h2>
      {props.children}
    </>
  )
}
export default PageTemplate
```

This works as expected! 

In the docs page: https://www.gatsbyjs.com/docs/mdx/programmatically-creating-pages/ they do a GraphQL query that breaks things when I add it. 

```
export const pageTemplateQuery = graphql`
  query PageTemplateQuery($id: String) {
    mdx(id: { eq: $id }) {
      id
      body
      frontmatter {
        title
      }
    }
  }
`
```
There is no id so it won't work. I guess since the mdx is already rendered I just need to put the mdx body in the form of `props.children`? Since I got rid of the changes in gatsby-node.js there is now no mdx data like it says on the docs page you linked: `Note: To query MDX content, it must be included in the node system using a source like the gatsby-source-filesystem plugin first. Instructions for sourcing content from somewhere like your /src/pages directory can be found on the plugin’s README.`

I now try doing just a metadata query for the page template like so:
```
  query PageTemplateQuery {
    site {
      siteMetadata {
        title
        description
        author
      }
    }
  }
```

This doesn't work and the console prints:
```
warning The GraphQL query in the non-page component "/<DIR>/src/templates/DefaultPageTemplate/DefaultPageTemplate.tsx" will not be run.
Exported queries are only executed for Page components. It's possible you're
trying to create pages in your gatsby-node.js and that's failing for some
reason.

If the failing component(s) is a regular component and not intended to be a page
component, you generally want to use a <StaticQuery> (https://gatsbyjs.org/docs/static-query)
instead of exporting a page query.

If you're more experienced with GraphQL, you can also export GraphQL
fragments from components and compose the fragments in the Page component
query and pass data down into the child component — https://graphql.org/learn/queries/#fragments
```

So it seems the page is already rendered, this template is not for the page but rather for layout, and you can only do Static GraphQL queries there because of that.

I will look more into the props passed into this layout component and I am guessing I will be able to (hopefully) everything needed and not require props. Just found the example here: https://github.com/gatsbyjs/gatsby/tree/master/examples/using-MDX and it seems that is what they do.

**Reply by nicolls1** [reply]
Update: I just saw that I have do do the query in the mdx file itself as on the page you linked: https://www.gatsbyjs.com/docs/mdx/writing-pages/#graphql-queries

This resolves my issue, thanks for the pointer and sorry for the disturbance!

**Reply by LekoArts** [reply]
No worries, happy to help! It's not necessarily your fault that you didn't find the doc 😅 We're reworking the structure, if you have feedback while you go through it we'd be happy to receive it here: https://github.com/gatsbyjs/gatsby/discussions/27856
